### ISSUE TITLE
MIP Callbacks

### ISSUE THREAD

**Issue by jajhall** [ERGO-Code/HiGHS#781]
Placeholder issue for the introduction of MIP callbacks, as discussed with @lgottwald 

Requested within https://github.com/coin-or/python-mip/issues/255

**Comment by odow** [comment]
Leaving a bread-crumb to https://jump.dev/JuMP.jl/stable/manual/callbacks/, which contains the documentation of the callbacks JuMP supports. However I think a good rule of thumb is "if in doubt, follow Gurobi."

**Comment by jajhall** [comment]
Since there are MIP callbacks for improving solutions, any solution and simple interrupt, and code for lazy constraints is WIP, this "prompt" issue seems unnecessary

**Comment by rlloretb** [comment]
Any news on callbacks for user cuts (different from lazy constraints)?

**Comment by jajhall** [comment]
What's the difference? 

Just the point at which the cuts are generated - in the cut generation phase, rather than when an improving integer feasible solution is obtained?

**Comment by rlloretb** [comment]
I agree that they are very similar. Technically, in Gurobi they are implemented differently due to incompatibilities with the preprocessing step (see precrush parameter). I was wondering if it is possible to add user cuts in HiGHs while the root node is being explored.

**Comment by jajhall** [comment]
OK thanks, that gives me a pointer to find out more.

**Comment by rlloretb** [comment]
So it is not possible to add cuts with the current callback implementation (v1.7.2)?

**Comment by jajhall** [comment]
User cuts, no

**Comment by rlloretb** [comment]
I see. Any plans to implement it within a year? It is a very powerful feature.

**Comment by jajhall** [comment]
There are plans - and money - to enhance the MIP solver over the coming 12 months

**Comment by Thell** [comment]
Hi Julian,
I've been wondering recently about HiGHS callbacks and am curious if a simple callback could be used that would allow a user to return an objective value modifier? Seeing that there are plans to work on callbacks this year I figured it'd be a good time to ask.

For example if the object is to maximize value minus cost but the cost is calculated out of HiGHS the callback would send the proposed solution and the user would simply send back the value... I guess you could say that it would essentially be setting a variable value via callback. Could that be a thing or ...?

**Comment by jajhall** [comment]
I can't see how we could allow this. HiGHS needs the explicit objective function to optimize.

**Comment by rschwarz** [comment]
> Since there are MIP callbacks for improving solutions [...]

To be sure, this means a callback that is triggered whenever a solution is found during the search, so that the user can log it, or stop the run, right?

What is **not** supported (or even planned?) is the possibility for the user to submit a solution from within a callback, right?

I tried calling `h.setSolution` from within a logging callback, and it only had an effect when this happened early enough (from the log, it seems that it needs to happen before presolve).

A possible use case for this would be primal heuristics, that apply some form of local search on each MIP solution found during the search.

I don't necessarily think that this is high priority, I just wanted to make sure I don't miss any undocumented functionality.

**Comment by jajhall** [comment]
> To be sure, this means a callback that is triggered whenever a solution is found during the search, so that the user can log it, or stop the run, right?

Yes

> What is **not** supported (or even planned?) is the possibility for the user to submit a solution from within a callback, right?

Yes. You seemed not to do any harm by calling `h.setSolution` from within a logging callback, but we really ought to have something to prevent calls to HiGHS methods in the callbacks, as behaviour is completely undefined! 

There are no plans to add this functionality, although I can see it would be useful for expert users.

**Comment by rschwarz** [comment]
I guess if we wanted to have a solver where **everything** can be customised, we can always use SCIP. So I can see that it makes sense to limit the functionality in HiGHS in order to control the complexity and maintenance effort 👍
